Allow disable operators from execution in frontend

zuozhiw opened the pull request:
This PR allows disable operators in the frontend from workflow execution.

Changes include:
 - add `disabledOperators` in workflow graph, add related functions and observables
 - add one button in `NavigationBar` that disable / enable all highlighted operators
 - change the operator background color in `JointUI` as operator is disabled/enabled
 - change `WorkflowValidation` logic to handle disabled operators
 - change `PersistWorkflow` logic to handle disabled operators
 - change `ExecuteWorkflow` logic to only submit enabled operators/links/breakpoints



![demo](https://user-images.githubusercontent.com/12578068/124578517-227fc880-de03-11eb-8917-95b8c92a31e7.gif)

Yicong-Huang commented on core/new-gui/src/app/common/type/workflow.ts:
```diff
@@ -14,14 +14,13 @@ import { Breakpoint, OperatorLink, OperatorPredicate, Point } from '../../worksp
+export interface WorkflowContent extends Readonly<{
   operators: OperatorPredicate[];
   operatorPositions: { [key: string]: Point };
+  disabledOperators: string[];
```
I am worried about the compatibility of this code. Remember we have workflows saved as JSON strings in prod DB. Changing the frontend representation of it in my understanding would read the `disabledOperators` as undefined. Would that cause some unexpected issues?

Yicong-Huang replied:
Another design might be to add a property on each `OperatorPredicate`, namely `disable: boolean`. In this way, we could do the same for `OperatorLink` in the future as well. 

what do you think?

zuozhiw replied:
changed `disabledOperators` to be optional so it's backwards compatible

zuozhiw replied:
Add a property to `OperatorPredicate` sounds a good idea, we can do the factor later because I don't want to spend too much time on this task,

zuozhiw replied:
changed the implementation to add an optional property in `OperatorPredicate`

Yicong-Huang commented on core/new-gui/src/app/workspace/component/navigation/navigation.component.ts:
```diff
@@ -111,6 +116,8 @@ export class NavigationComponent implements OnInit {
 
     this.registerWorkflowMetadataDisplayRefresh();
+
+    this.handleDisableOperatorIconChange();
```
maybe name it `registerDisableOperatorIconChange`? so that you register a handler for an event.

zuozhiw replied:
in many places we use `handleXXX` for this kind of functions

Yicong-Huang commented on core/new-gui/src/app/workspace/component/navigation/navigation.component.ts:
```diff
@@ -325,6 +332,19 @@ export class NavigationComponent implements OnInit {
     }
   }
 
+  public onClickDisableOperators(): void {
```
it will be good to include some docstring and comments. 

What happens if the highlights contain both disabled and undisabled operators and I click "disable"?

Yicong-Huang commented on core/new-gui/src/app/workspace/component/navigation/navigation.component.ts:
```diff
@@ -371,4 +391,36 @@ export class NavigationComponent implements OnInit {
 
+  handleDisableOperatorIconChange() {
+    Observable.merge(
+      this.workflowActionService.getJointGraphWrapper().getJointOperatorHighlightStream(),
+      this.workflowActionService.getJointGraphWrapper().getJointOperatorUnhighlightStream(),
+      this.workflowActionService.getJointGraphWrapper().getJointGroupHighlightStream(),
+      this.workflowActionService.getJointGraphWrapper().getJointGroupUnhighlightStream(),
+      this.workflowActionService.getTexeraGraph().getDisabledOperatorsChangedStream(),
```
as this block of code is showing many places, how about we do a merged observable inside WorkflowActionService to reduce duplication of code?

zuozhiw replied:
this place is different than other places

the reason that you might feel it's a lot is because of there are so many highlight events, we can refactor them in to one single observable later

Yicong-Huang commented on core/new-gui/src/app/workspace/component/navigation/navigation.component.ts:
```diff
@@ -371,4 +391,36 @@ export class NavigationComponent implements OnInit {
+
+  calculateIsDisableOperator(): boolean {
+    const allDisabled = this.effectivelyHighlightedOperators().every(
+      op => this.workflowActionService.getTexeraGraph().isOperatorDisabled(op));
```
if the disabled property is on the operator itself, isn't that easier to check if it is disabled or not?

zuozhiw replied:
I don't see a difference here because the api will be the same, the highlighted operator is an operator ID instead of the operator object

Yicong-Huang commented on core/new-gui/src/app/workspace/component/navigation/navigation.component.ts:
```diff
@@ -371,4 +391,36 @@ export class NavigationComponent implements OnInit {
+    });
+  }
+
+  calculateIsDisableOperator(): boolean {
```
it will be good to give some docstring/comments. The function name is somewhat confusing.

zuozhiw replied:
added

Yicong-Huang commented on core/new-gui/src/app/workspace/component/result-panel/result-panel.component.ts:
```diff
@@ -105,7 +105,9 @@ export class ResultPanelComponent {
         const sinkOperators = this.workflowActionService.getTexeraGraph().getAllOperators()
           .filter(op => op.operatorType.toLowerCase().includes('sink'));
-        if (sinkOperators.length > 0) {
+        if (sinkOperators.length > 0 && ! this.resultPanelOperatorID) {
```
as I read here, this came into my mind:

what happens if I disable a view result operator after the workflow is successfully executed, and I want to see the results? Can I still view results?

zuozhiw replied:
I believe you can

Yicong-Huang commented on core/new-gui/src/app/workspace/service/workflow-graph/model/workflow-graph.ts:
```diff
@@ -69,10 +66,47 @@ export class WorkflowGraph {
     if (!operator) {
       throw new Error(`operator with ID ${operatorID} doesn't exist`);
     }
+    if (this.isOperatorDisabled(operatorID)) {
+      this.enableOperator(operatorID);
+    }
```
I see the intention of enabling an operator so that you can delete it correctly. 
But I wonder if there are better ways to do this, several reasons:
1. Enable an operator would trigger a lot of events, such as workflow persist, validation, schema propagation, etc. All those are unnecessary overheads.
2. Would the user "see" any effect?

zuozhiw replied:
changed to just delete from the map without trigger the event

Yicong-Huang commented on core/new-gui/src/app/workspace/service/workflow-status/workflow-status.service.ts:
```diff
@@ -31,19 +31,6 @@ export class WorkflowStatusService {
       this.statusSubject.next(event.operatorStatistics);
     });
 
-    this.executeWorkflowService.getExecutionStateStream().subscribe(event => {
-      if (event.current.state === ExecutionState.WaitingToRun) {
-        const initialStatistics: Record<string, OperatorStatistics> = {};
-        this.workflowActionService.getTexeraGraph().getAllOperators().forEach(op => {
-          initialStatistics[op.operatorID] = {
-            operatorState: OperatorState.Initializing,
-            aggregatedInputRowCount: 0,
-            aggregatedOutputRowCount: 0,
-          };
-        });
-        this.statusSubject.next(initialStatistics);
-      }
-    });
```
what is this change about?

zuozhiw replied:
frontend changes all operator's state to `initializing` when user clicks run, including the disabled operators.
originally I thought to exclude the disabled operators, but then I just decided this frontend trick is useless and might be confusing, so I just removed it.

Yicong-Huang reviewed (changes requested):
I love this feature! Left some comments, please take care of them.

zuozhiw commented:
@Yicong-Huang thanks for the review, updated the code based on your feedback and added some test cases, please check again, also please test it again, thanks.

Yicong-Huang reviewed (approved):
I've tested the functionality, some comments:
1. It is now backward compatible with the persisted workflows. Thanks!
2. When having a workflow but with all operators disabled, the "Run" button is not deactivated. This might be another issue that's unrelated to this PR: an empty workflow is having "Run" button activated.
3. When disabling some operator in a workflow, schema propagation is NOT retriggered, thus the enabled operators still see the schema as if the target operator is not disabled. Also, when a disabled operator is enabled, schema propagation is again not triggered, resulting in a broken workflow. I think it is a minor issue though: the user can always refresh to trigger schema propagation, but better to do it automatically when disable/enable the operator.
4. Undo/redo is not in place or functioning correctly. We either fix it or ignore it for now.
